File: packages/python/codecontext_python/checker.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Iterable

from .parser import find_project_root, parse_source


@dataclass(slots=True)
class CheckMessage:
    file: str
    line: int
    message: str


def check_paths(
    paths: Iterable[str | Path],
    *,
    context_dir: str = "docs/context",
    check_refs: bool = True,
) -> list[CheckMessage]:
    messages: list[CheckMessage] = []

    for file_path in _iter_python_files(paths):
        source = file_path.read_text(encoding="utf-8")
        project_root = find_project_root(file_path.parent) or file_path.parent
        result = parse_source(
            source,
            project_root,
            context_dir,
            check_refs=check_refs,
        )

        unresolved_lines = {tag.id: tag.line for tag in result.tags if tag.id}

        for error in result.errors:
            line = 1
            if error.startswith("Unresolved context reference: "):
                ref = error.removeprefix('Unresolved context reference: "').removesuffix('"')
                line = unresolved_lines.get(ref, 1)
            elif error.startswith("Malformed @context tag: "):
                line = _first_context_line(source)
            else:
                for tag in result.tags:
                    if tag.line:
                        line = tag.line
                        break
            messages.append(CheckMessage(str(file_path), line, error))

    return messages


def _iter_python_files(paths: Iterable[str | Path]) -> list[Path]:
    files: list[Path] = []
    for raw_path in paths:
        path = Path(raw_path)
        if path.is_dir():
            files.extend(sorted(candidate for candidate in path.rglob("*.py") if candidate.is_file()))
        elif path.suffix == ".py" and path.is_file():
            files.append(path)
    seen: set[Path] = set()
    unique: list[Path] = []
    for file_path in files:
        resolved = file_path.resolve()
        if resolved in seen:
            continue
        seen.add(resolved)
        unique.append(file_path)
    return unique


def _first_context_line(source: str) -> int:
    for line_number, line in enumerate(source.splitlines(), start=1):
        if "@context" in line:
            return line_number
    return 1

File: packages/ruby/lib/codecontext/ruby/checker.rb
# frozen_string_literal: true

require "pathname"
require_relative "parser"

module CodeContext
  module Ruby
    CheckMessage = Struct.new(:file, :line, :message, keyword_init: true)

    module_function

    def check_paths(paths, context_dir: "docs/context", check_refs: true)
      messages = []

      iter_ruby_files(paths).each do |file_path|
        source = file_path.read
        project_root = find_project_root(file_path.dirname) || file_path.dirname
        result = parse_source(source, project_root, context_dir: context_dir, check_refs: check_refs)
        unresolved_lines = result.tags.each_with_object({}) { |tag, acc| acc[tag.id] = tag.line if tag.id }

        result.errors.each do |error|
          line = line_for_error(error, source, result.tags, unresolved_lines)
          messages << CheckMessage.new(file: file_path.to_s, line: line, message: error)
        end
      end

      messages
    end

    def iter_ruby_files(paths)
      files = paths.flat_map do |raw_path|
        path = Pathname(raw_path)
        if path.directory?
          Dir.glob(path.join("**/*.rb").to_s).sort.map { |candidate| Pathname(candidate) }.select(&:file?)
        elsif path.file? && path.extname == ".rb"
          [path]
        else
          []
        end
      end

      files.uniq(&:realpath)
    end

    def line_for_error(error, source, tags, unresolved_lines)
      return unresolved_lines.fetch(error.delete_prefix('Unresolved context reference: "').delete_suffix('"'), 1) if error.start_with?("Unresolved context reference: ")
      return first_context_line(source) if error.start_with?("Malformed @context tag: ")

      tags.first&.line || 1
    end

    def first_context_line(source)
      source.each_line.with_index(1) do |line, line_number|
        return line_number if line.include?("@context")
      end

      1
    end
  end
end
